libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -216,17 +209,34 @@ impl NotificationHandler {
         }
     }
 
-    pub fn unsubscribe(&self) -> Result<(), ConnectionManagerError> {
-        let message = CmMessage::UnSubscribe(self.id.clone());
-        match self.sender.send(message) {
-            Ok(()) => Ok(()),
-            Err(_) => Err(ConnectionManagerError::SendMessageError(
-                "Unsubscribe request timed out".into(),
+    #[cfg(feature = "connection-manager-notifier-try-next")]
+    pub fn try_listen(
+        &self,
+    ) -> Result<Option<ConnectionManagerNotification>, ConnectionManagerError> {
+        match self.recv.try_recv() {
+            Ok(notifications) => Ok(Some(notifications)),
+            Err(TryRecvError::Empty) => Ok(None),
+            Err(TryRecvError::Disconnected) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
             )),
         }
     }
 }
 
+impl Iterator for Notifier {
+    type Item = ConnectionManagerNotification;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.recv.recv() {
+            Ok(notification) => Some(notification),
+            Err(_) => {
```

Comment:
I think the Notifier should explicitly recv a end-of-iteration marker so that if we get an err on receive it is unexpected and means there was a coding error.

Reply:
I'm not entirely sure how this would work, or if it's necessary. Again, `recv` can only "fail" if the connection manger has gone away either by shutting down gracefully or by panicking. Regardless, from the perspective of the iterator, it's done iterating. Adding an explicit marker seems redundant since the "marker" is the sender end going out of scope.

It's worth noting there is a `FatalError` notification that gets sent out if the connection manager is forced to shutdown unexpectedly, so the subscriber would be notified if something happened to connection manager and would be able to act accordingly.

Reply:
Ok, suggest removing the debug! then (completely uninteresting from a debugging standpoint, as you can't tell if it is normal operation or not) and adding a comment on that line reading "This is expected if the other side shuts down before this end." (re-used from Transact code).